pybamm/discretisations/base_discretisation.py

```diff
@@ -0,0 +1,306 @@
+            rhs[variable] = self.process_symbol(
+                equation, variable.domain, y_slices, boundary_conditions
+            )
+
+        # Concatenate and evaluate right-hand sides
+        self._concatenated_rhs = pybamm.NumpyConcatenation(*rhs.values())
+
+        def dydt(t, y):
+            return self._concatenated_rhs.evaluate(t, y)
+
+        return dydt
+
+    def process_symbol(self, symbol, domain, y_slices=None, boundary_conditions={}):
+        """Discretise operators in model equations.
+
+        Parameters
+        ----------
+        symbol : :class:`pybamm.expression_tree.symbol.Symbol` (or subclass) instance
+            Symbol to discretise
+        y_slices : dict of {Variable: slice}
+            The slices to assign to StateVectors when discretising a variable
+
+        Returns
+        -------
+        :class:`pybamm.expression_tree.symbol.Symbol` (or subclass) instance
+            Discretised symbol
+
+        """
+        if isinstance(symbol, pybamm.Gradient):
+            return self.gradient(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+
+        if isinstance(symbol, pybamm.Divergence):
+            return self.divergence(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+
+        elif isinstance(symbol, pybamm.BinaryOperator):
+            new_left = self.process_symbol(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+            new_right = self.process_symbol(
+                symbol.children[1], domain, y_slices, boundary_conditions
+            )
+            return symbol.__class__(new_left, new_right)
+
+        elif isinstance(symbol, pybamm.UnaryOperator):
+            new_child = self.process_symbol(
+                symbol.children[0], domain, y_slices, boundary_conditions
+            )
+            return symbol.__class__(new_child)
+
+        elif isinstance(symbol, pybamm.Variable):
+            return pybamm.StateVector(y_slices[symbol.id])
+
+        elif isinstance(symbol, pybamm.Scalar):
+            return pybamm.Scalar(symbol.value)
+
+        else:
+            raise TypeError("""Cannot discretise {!r}""".format(symbol))
+
+    def gradient(self, symbol, domain, y_slices, boundary_conditions):
+        """How to discretise gradient operators.
+
+        Parameters
+        ----------
+        symbol : :class:`Symbol` (or subclass)
+            The symbol (typically a variable) of which to take the gradient
+        domain : list
+            The domain(s) in which to take the gradient
+        y_slices : slice
+            The slice to assign to StateVector when discretising a variable
+        boundary_conditions : dict
+            The boundary conditions of the model ({symbol.id: (left bc, right bc)})
+
+        """
+        # Check that boundary condition keys are hashes (ids)
+        for key in boundary_conditions.keys():
```

Comment:
I don't understand what "overrirdden by the subclass" means - I put this here because passing symbols (not ids) as the keys of `boundary_conditions` might lead to the required ids not being found in the keys `boundary_conditions` for hard-to-understand reasons (see below comments on ids)

Reply:
but the derived classes of `BaseDiscretisation` (e.g. `MatrixVectorDiscretisation`) define their own `gradient` functions, so this function will never be used (and if by some chance it does, it will raise an `NotImplementedError` exception anyway).

Reply:
Oh I understand now, good spot